keps/sig-instrumentation/383-new-event-api-ga-graduation/README.md

```diff
@@ -439,6 +449,198 @@ List Events from the NamespaceSystem with field selector `reportingController =
 
 List all Event with field selector `regarding.name = podName, regarding.namespace = podNamespace`, and `related.name = podName, related.namespace = podNamespace`. You need to join results outside of the kubernetes API.
 
+## Production Readiness Review Questionnaire
+
+### Feature enablement and rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+  - [ ] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name:
+    - Components depending on the feature gate:
+  - [x] Other
+    - Describe the mechanism:
+
+      The new Event API can be enabled / disabled by switching to use the new / old Event proto and libs.
```

Comment:
Actually, let me pull in others opinions.

@johnbelamaric - given my explanation above (ignoring for a moment a fact that it was a bad decision and we shouldn't do this that way), I'm willing towards saying that "rollback is not supported". Or should we interpret it as "if you rollback to the release where component X wasn't migrated to use the new API, it will be rolled back".

[FTR - currently only scheduler is migrated to the new API]

Reply:
A few things. I think, "rollback supported" is probably a bad term. Using "rollback supported" to describe the ability to *disable* the feature (i.e, roll back the *enablement* as described below) without rolling back the *version* I think is a little confusing. We probably should call it "disable-supported".

In this case, it *sounds* like you are saying, the way to disable this is not via a feature gate but instead by disabling the API, in which case the library is smart enough to use the prior API. IIUC, this causes no data loss on existing events created with the new API, and presumably we round-trip those events so you'll still be able to query events created with the new API after disabling it, by using the old API. If that's correct, then I would say "rollback supported" is true. It's not feature gate based, but it is possible and works.

Reply:
> We probably should call it "disable-supported".

+1

> If that's correct, then I would say "rollback supported" is true. It's not feature gate based, but it is possible and works.

Yes - what yo wrote is the right summary.
Fair enough - I think that makes sense to me.

Reply:
@liggitt - IIRC we discussed a while ago fallbacking to core. once we move move to GA I think we should still fallback for safety. might be worth considering for other APIs (if we're not already) ?

Reply:
It depends on the use. the event recorder is used by a *broad* range of clients, many out of tree, with potentially different skew support statements than kube-scheduler/kube-controller-manager, etc.

The broader the use, the more we should accomodate running against a range of server versions, so fallback for event recorder makes sense to me.

Reply:
> so fallback for event recorder makes sense to me.

Agree (at least for the next couple releases).

Also - currently this logic is purely baked into scheduler:
https://github.com/kubernetes/kubernetes/blob/96e13de777a9eb57f87889072b68ac40467209ac/cmd/kube-scheduler/app/server.go#L281
We should probably make that part of the library.

Reply:
I updated the description of the mechanism, not sure if I got it correct. Let me know if it needs any changes.